File: src/udp_tts/chunker.py
"""Split book text into TTS-friendly chunks.

TTS quality and latency are best when each request is a small, self-contained
piece of text that ends on a natural boundary. This module splits text into
sentences and greedily packs them into chunks up to ``max_chars`` without ever
breaking mid-sentence -- unless a single sentence is itself longer than the
limit, in which case it is split at clause or word boundaries as a fallback.

No NLP dependencies: sentence segmentation is a pragmatic regex that handles the
common cases (terminal punctuation, quotes/brackets, common abbreviations).
"""

import re
from dataclasses import dataclass
from typing import List, Optional

DEFAULT_MAX_CHARS = 400

# Abbreviations that end in a period but rarely end a sentence.
_ABBREVIATIONS = {
    "mr", "mrs", "ms", "dr", "prof", "sr", "jr", "st", "vs", "etc", "e.g",
    "i.e", "fig", "no", "vol", "pp", "al", "inc", "ltd", "co",
}

# A sentence boundary: terminal punctuation, optional closing quote/bracket,
# then whitespace, then something that looks like the start of a new sentence.
_BOUNDARY = re.compile(
    r"""(?<=[.!?])["'”’)\]]?\s+(?=["'“‘(\[]?[A-Z0-9])""",
    re.VERBOSE,
)


@dataclass
class Chunk:
    """One unit of text to synthesize, with its place in the book."""
    text: str
    index: int                      # global 0-based chunk index
    chapter_index: int
    chapter_title: Optional[str]
    is_chapter_start: bool


def split_sentences(text: str) -> List[str]:
    """Split a block of text into sentences (best-effort, dependency-free)."""
    text = text.strip()
    if not text:
        return []
    # Split paragraphs first so we never merge across blank lines.
    sentences: List[str] = []
    for paragraph in re.split(r"\n\s*\n", text):
        paragraph = " ".join(paragraph.split())
        if not paragraph:
            continue
        pieces = _BOUNDARY.split(paragraph)
        sentences.extend(_merge_abbreviations(pieces))
    return [s.strip() for s in sentences if s.strip()]


def _merge_abbreviations(pieces: List[str]) -> List[str]:
    """Re-join splits that occurred right after a common abbreviation."""
    merged: List[str] = []
    for piece in pieces:
        if merged:
            last = merged[-1]
            tail = last.rsplit(" ", 1)[-1].rstrip(".").lower()
            if tail in _ABBREVIATIONS:
                merged[-1] = last + " " + piece
                continue
        merged.append(piece)
    return merged


def _split_long_sentence(sentence: str, max_chars: int) -> List[str]:
    """Break an over-long sentence at clause, then word, boundaries."""
    if len(sentence) <= max_chars:
        return [sentence]
    # Prefer clause boundaries (comma, semicolon, colon, dash).
    parts = re.split(r"(?<=[,;:—])\s+", sentence)
    out: List[str] = []
    buf = ""
    for part in parts:
        candidate = (buf + " " + part).strip() if buf else part
        if len(candidate) <= max_chars:
            buf = candidate
        else:
            if buf:
                out.append(buf)
            if len(part) <= max_chars:
                buf = part
            else:
                out.extend(_split_by_words(part, max_chars))
                buf = ""
    if buf:
        out.append(buf)
    return out


def _split_by_words(text: str, max_chars: int) -> List[str]:
    out, buf = [], ""
    for word in text.split():
        candidate = (buf + " " + word).strip() if buf else word
        if len(candidate) <= max_chars:
            buf = candidate
        else:
            if buf:
                out.append(buf)
            buf = word  # a single word longer than max_chars is left intact
    if buf:
        out.append(buf)
    return out


def chunk_text(text: str, max_chars: int = DEFAULT_MAX_CHARS) -> List[str]:
    """Pack the sentences of ``text`` into chunks no longer than ``max_chars``."""
    if max_chars < 1:
        raise ValueError("max_chars must be >= 1")
    chunks: List[str] = []
    buf = ""
    for sentence in split_sentences(text):
        for piece in _split_long_sentence(sentence, max_chars):
            candidate = (buf + " " + piece).strip() if buf else piece
            if len(candidate) <= max_chars:
                buf = candidate
            else:
                if buf:
                    chunks.append(buf)
                buf = piece
    if buf:
        chunks.append(buf)
    return chunks


def chunk_book(book, max_chars: int = DEFAULT_MAX_CHARS) -> List[Chunk]:
    """Chunk every chapter of a :class:`udp_tts.book.Book` in reading order."""
    chunks: List[Chunk] = []
    index = 0
    for chapter_index, chapter in enumerate(book.chapters):
        texts = chunk_text(chapter.text, max_chars)
        for i, text in enumerate(texts):
            chunks.append(Chunk(
                text=text,
                index=index,
                chapter_index=chapter_index,
                chapter_title=chapter.title,
                is_chapter_start=(i == 0),
            ))
            index += 1
    return chunks

File: macos-client/Sources/UDPTTSCore/Chunker.swift
import Foundation

/// One unit of text to synthesize, with its place in the book.
public struct BookChunk: Equatable {
    public let text: String
    public let index: Int
    public let chapterIndex: Int
    public let chapterTitle: String?
    public let isChapterStart: Bool
}

/// Splits book text into TTS-friendly chunks. Swift port of `chunker.py`:
/// segment into sentences, then greedily pack sentences into chunks up to
/// `maxChars` without breaking mid-sentence (over-long sentences fall back to
/// clause, then word, boundaries).
public enum Chunker {
    public static let defaultMaxChars = 400

    private static let abbreviations: Set<String> = [
        "mr", "mrs", "ms", "dr", "prof", "sr", "jr", "st", "vs", "etc", "e.g",
        "i.e", "fig", "no", "vol", "pp", "al", "inc", "ltd", "co",
    ]

    // MARK: Sentence segmentation

    public static func splitSentences(_ text: String) -> [String] {
        var sentences: [String] = []
        // Split into paragraphs on blank lines so we never merge across them.
        let paragraphs = text.components(separatedBy: blankLineRegex)
        for para in paragraphs {
            let collapsed = para.split(whereSeparator: { $0.isWhitespace })
                .joined(separator: " ")
            if collapsed.isEmpty { continue }
            sentences.append(contentsOf: segment(collapsed))
        }
        return sentences
    }

    /// Manual scanner: break after .!? (+ optional closing quote) when followed
    /// by whitespace and a capital/digit/quote, unless the preceding token is a
    /// known abbreviation.
    private static func segment(_ paragraph: String) -> [String] {
        let chars = Array(paragraph)
        var sentences: [String] = []
        var start = 0
        var i = 0
        let terminators: Set<Character> = [".", "!", "?"]
        let closers: Set<Character> = ["\"", "'", "”", "’", ")", "]"]
        let openers: Set<Character> = ["\"", "'", "“", "‘", "(", "["]

        while i < chars.count {
            if terminators.contains(chars[i]) {
                var j = i + 1
                while j < chars.count && closers.contains(chars[j]) { j += 1 }
                // require whitespace next
                guard j < chars.count, chars[j].isWhitespace else { i += 1; continue }
                var k = j
                while k < chars.count && chars[k].isWhitespace { k += 1 }
                guard k < chars.count else { break }
                let next = chars[k]
                let startsNewSentence = next.isUppercase || next.isNumber || openers.contains(next)
                if startsNewSentence && !endsWithAbbreviation(chars, from: start, to: i) {
                    let sentence = String(chars[start..<j]).trimmingCharacters(in: .whitespaces)
                    if !sentence.isEmpty { sentences.append(sentence) }
                    start = k
                    i = k
                    continue
                }
            }
            i += 1
        }
        let tail = String(chars[start..<chars.count]).trimmingCharacters(in: .whitespaces)
        if !tail.isEmpty { sentences.append(tail) }
        return sentences
    }

    private static func endsWithAbbreviation(_ chars: [Character], from: Int, to: Int) -> Bool {
        // Last whitespace-delimited token in chars[from..<to], minus trailing dots.
        var s = to
        while s > from && !chars[s - 1].isWhitespace { s -= 1 }
        let token = String(chars[s..<to]).lowercased()
            .trimmingCharacters(in: CharacterSet(charactersIn: "."))
        return abbreviations.contains(token)
    }

    // MARK: Packing

    public static func chunkText(_ text: String, maxChars: Int = defaultMaxChars) -> [String] {
        precondition(maxChars >= 1)
        var chunks: [String] = []
        var buf = ""
        for sentence in splitSentences(text) {
            for piece in splitLongSentence(sentence, maxChars: maxChars) {
                let candidate = buf.isEmpty ? piece : buf + " " + piece
                if candidate.count <= maxChars {
                    buf = candidate
                } else {
                    if !buf.isEmpty { chunks.append(buf) }
                    buf = piece
                }
            }
        }
        if !buf.isEmpty { chunks.append(buf) }
        return chunks
    }

    private static func splitLongSentence(_ sentence: String, maxChars: Int) -> [String] {
        if sentence.count <= maxChars { return [sentence] }
        let parts = sentence.components(separatedBy: clauseRegex)
        var out: [String] = []
        var buf = ""
        for part in parts {
            let candidate = buf.isEmpty ? part : buf + " " + part
            if candidate.count <= maxChars {
                buf = candidate
            } else {
                if !buf.isEmpty { out.append(buf) }
                if part.count <= maxChars {
                    buf = part
                } else {
                    out.append(contentsOf: splitByWords(part, maxChars: maxChars))
                    buf = ""
                }
            }
        }
        if !buf.isEmpty { out.append(buf) }
        return out
    }

    private static func splitByWords(_ text: String, maxChars: Int) -> [String] {
        var out: [String] = []
        var buf = ""
        for word in text.split(separator: " ") {
            let candidate = buf.isEmpty ? String(word) : buf + " " + word
            if candidate.count <= maxChars {
                buf = candidate
            } else {
                if !buf.isEmpty { out.append(buf) }
                buf = String(word)  // a single over-long word is left intact
            }
        }
        if !buf.isEmpty { out.append(buf) }
        return out
    }

    public static func chunkBook(_ book: Book, maxChars: Int = defaultMaxChars) -> [BookChunk] {
        var chunks: [BookChunk] = []
        var index = 0
        for (chapterIndex, chapter) in book.chapters.enumerated() {
            let texts = chunkText(chapter.text, maxChars: maxChars)
            for (i, text) in texts.enumerated() {
                chunks.append(BookChunk(
                    text: text, index: index, chapterIndex: chapterIndex,
                    chapterTitle: chapter.title, isChapterStart: i == 0))
                index += 1
            }
        }
        return chunks
    }

    // MARK: Regexes (compiled once)

    private static let blankLineRegex = try! NSRegularExpression(pattern: "\\n\\s*\\n")
    private static let clauseRegex = try! NSRegularExpression(pattern: "(?<=[,;:—])\\s+")
}

private extension String {
    /// Split on each match of `regex`, returning the pieces between matches.
    func components(separatedBy regex: NSRegularExpression) -> [String] {
        let ns = self as NSString
        let matches = regex.matches(in: self, range: NSRange(location: 0, length: ns.length))
        var pieces: [String] = []
        var last = 0
        for m in matches {
            pieces.append(ns.substring(with: NSRange(location: last, length: m.range.location - last)))
            last = m.range.location + m.range.length
        }
        pieces.append(ns.substring(from: last))
        return pieces
    }
}
